fix(zone.js): support addEventListener with signal option.

JiaLiPassion opened the pull request:
Close #49591

```
const ac = new AbortController();
addEventListener(eventName, handler, {signal: ac.signal);`
ac.abort();
```

Currently `zone.js` doesn't support the `signal` option, this PR allows the user to use AbortContoller to remove the event listener.

JiaLiPassion committed:
fix(zone.js): support addEventListener with signal option.

Close #49591

```
const ac = new AbortController();
addEventListener(eventName, handler, {signal: ac.signal);`
ac.abort();
```

Currently `zone.js` doesn't support the `signal` option, this PR allows
the user to use AbortContoller to remove the event listener.

DavidANeil commented on packages/zone.js/test/browser/browser.spec.ts:
```diff
@@ -1763,6 +1763,32 @@ describe('Zone', function() {
            expect(logs).toEqual(['click2']);
          });
 
+      it('should support remove event listeners via AbortController', function() {
```
I think it would be worthwhile to add a new test where the signal is aborted _before_ it is passed in to `addEventListener`.
That should result in the listener being immediately removed (or never attached), but I am doubtful that is the case with this implementation.

JiaLiPassion replied:
@DavidANeil , Thanks for the review, yeah, you are right, I have added the case also handle the same signal cancel multiple event listeners case.

DavidANeil reviewed (approved):
I think this design isn't satisfactory (if this were merged, I would probably patch it back to my "just backoff entirely when a signal is passed in" mode).

But I think it is better than the status quo, so I am approving for the sake of incremental improvement.

DavidANeil commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -468,6 +473,12 @@ export function patchEventTarget(
 
+        if (signal && signal.abortController) {
+          const tasks = signal.abortController.tasks || [];
+          tasks.push(task);
```
It probably also leaks listeners added with `once: true`, which is a much more likely scenario.

JiaLiPassion replied:
@DavidANeil 
Thanks for pointing out this case, I added the test case and the logic to handle this case.

```
const aLargeObject = createLargeObject();
const holdsLotsOfStuffInMemory = function() {
    console.log(aLargeObject);
}
const ac = new AbortController();
target.addEventListener('someEvent', holdsLotsOfStuffInMemory, {signal: ac.signal});
target.removeEventListener('someEvent', holdsLotsOfStuffInMemory); // (signal isn't required to remove a listener)
return ac;

```
after `removeEventListener`, the `task` will be removed from `ac.tasks` array.

alan-agius4 replied:
Maybe, we could store a weak reference to the task instead of a strong one, which would cause the object to be GC'd for the `once` use-case. 

```ts
tasks.push(new WeakRef(task));
```

JiaLiPassion commented on packages/zone.js/test/browser/browser.spec.ts:
```diff
@@ -1763,6 +1763,78 @@ describe('Zone', function() {
+
+               button.addEventListener('click', function() {
+                 logs.push('click1');
+               }, {signal: AbortSignal.timeout(1)});
```
> I think this design isn't satisfactory (if this were merged, I would probably patch it back to my "just backoff entirely when a signal is passed in" mode).

could you explain more about this? I am not sure which part is not compatible with the DOM spec.

Also I tested with vanilla js.

```
               button.addEventListener('click', function() {
                 logs.push('click1');
               }, {signal: AbortSignal.timeout(1)});
```
will addEventListener successfully event after 1ms the signal is timeout and aborted, the event listener is not removed.

DavidANeil replied:
My testing is showing the listener _is_ removed in this case with vanilla js.

JiaLiPassion commented on packages/zone.js/lib/common/abort-controller.ts:
```diff
@@ -0,0 +1,31 @@
+    api.patchMethod(
+        AbortController.prototype, 'abort', (delegate: Function) => (self: any, args: any) => {
+          if (Array.isArray(self.tasks)) {
+            self.tasks.forEach((task: Task) => task.zone.cancelTask(task));
```
Thanks, updated.

DavidANeil commented:
I believe you've fixed the memory leaks, so this is looking good.

I'm still not sure if something can be done about `AbortSignal.timeout`, though it can be done later in another PR.

physedo commented:
@AndrewKushnir  ?  Is this fixed?

jeripeierSBB commented:
@JiaLiPassion @DavidANeil Would there be anything where I can support this PR? 
We would like to use a lib which uses AbortController in their implementation and we cannot use it together with an Angular application (as long as it uses zone.js).

DavidANeil commented on packages/zone.js/lib/common/abort-controller.ts:
```diff
@@ -0,0 +1,32 @@
+Zone.__load_patch('AbortController', (global: any, Zone: ZoneType, api: _ZonePrivate) => {
+  const AbortController = global['AbortController'];
+  if (AbortController) {
+    global['AbortController'] = function() {
```
Would this break `instanceof AbortController`?

kyubisation commented:
@JiaLiPassion @DavidANeil As @jeripeierSBB mentioned; Is there any way to help? With the wider usage of AbortController in the web, the lack of support in zone.js is an intransparent error for consumers, that have no idea why specific code fails.

JiaLiPassion commented:
I will check this one this week, thanks.

kyubisation commented:
@JiaLiPassion Thank you very much. I apologize for being pushy.

JiaLiPassion committed:
fix(zone.js): handle fetch with AbortSignal

fetch support AbortSignal, zone.js schedules a macroTask when fetch()

```
fetch(..., {signal: abortSignal});
```

we should also be able to cancel fetch with `zoneTask.cancel` call.
So this commit create an internal AbortSignal to handle
`zoneTask.cancel()` call and also delegate the `options.signal` from the
user code.

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -465,9 +480,31 @@ export function patchEventTarget(
+        if (signal && typeof signal === 'object') {
+          // after task is scheduled, we need to store the signal back to task.options
+          taskData.options.signal = signal;
+          const tasks = signal.tasks || [];
+          tasks.push(task);
+          signal.tasks = tasks;
```
```suggestion
          (signal.tasks ??= []).push(task);
```

JiaLiPassion replied:
Thanks, this is much better.

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -465,9 +480,31 @@ export function patchEventTarget(
+          if (!signal[Zone.__symbol__('abortListener')]) {
+            signal[Zone.__symbol__('abortListener')] = true;
+            nativeListener.call(signal, 'abort', function() {
+              const sTasks = signal.tasks.slice();
```
Why we need to do a shallow copy?

JiaLiPassion replied:
The reason is the same signal can be used on multiple event listeners, so make a shallow copy will avoid different event listeners impact each others.
But we don't need this `tasks` array at all, I just used the `closure` to cache the `task`.

DavidANeil commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -465,9 +480,31 @@ export function patchEventTarget(
+          signal.tasks = tasks;
+          if (!signal[Zone.__symbol__('abortListener')]) {
+            signal[Zone.__symbol__('abortListener')] = true;
+            nativeListener.call(signal, 'abort', function() {
```
I believe this should be `{once: true}`.
`'abort'` should only ever fire once, so I just always use `once: true`, though it is possible the useragent already cleans them up.

JiaLiPassion replied:
Sure, I added the `once: true` option, thanks.

DavidANeil commented on packages/zone.js/lib/common/fetch.ts:
```diff
@@ -26,31 +26,22 @@ Zone.__load_patch('fetch', (global: any, Zone: ZoneType, api: _ZonePrivate) => {
+    if (signal) {
+      const nativeAddEventListener =
+          signal[Zone.__symbol__('addEventListener')] || signal.addEventListener;
+      nativeAddEventListener.call(signal, 'abort', function() {
```
Also with `{once: true}`.

JiaLiPassion replied:
Sure, I added the `once: true` option, thanks.

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -293,6 +293,16 @@ export function patchEventTarget(
+      const signal = task?.options?.signal;
+      if (typeof signal === 'object' && signal?.tasks) {
+        const abortTasks = signal.tasks;
+        for (let i = 0; i < abortTasks?.length || 0; i++) {
+          if (abortTasks[i] === task) {
+            abortTasks.splice(i, 1);
+            break;
+          }
+        }
```
```suggestion
        const taskIdx = abortTasks.indexOf(task);
        if (taskIdx !== -1) {
            abortTasks.splice(taskIdx, 1);
        }
```

JiaLiPassion replied:
Thanks, we don't need the `tasks` array, so I just removed the part of code.

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -465,9 +480,31 @@ export function patchEventTarget(
           (data as any).taskData = taskData;
         }
 
+        if (signal && typeof signal === 'object') {
+          // if addEventListener with signal options, we don't pass it to
+          // native addEventListener, instead we keep the signal setting
+          // and handle ourselves.
+          taskData.options.signal = undefined;
+        }
```
Can't this logic be combined with that of line 492+?

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -394,6 +404,11 @@ export function patchEventTarget(
         const passive =
             passiveSupported && !!passiveEvents && passiveEvents.indexOf(eventName) !== -1;
         const options = buildEventListenerOptions(arguments[2], passive);
+        const signal = typeof options === 'object' && options?.signal;
```
Consider updating the below to 
```suggestion
       const signal = options && typeof options === 'object' && options.signal && typeof options.signal === 'object' ? options.signal : undefined;
```

To avoid the repetitive checks... 
```ts
if (typeof signal === 'object' && signal?.aborted) 
if (signal && typeof signal === 'object')
```

JiaLiPassion replied:
Got it, updated, thanks.

alan-agius4 reviewed (approved):
LGTM

alan-agius4 commented:
[TGP](https://test.corp.google.com/OCL:590580743:BASE:590580467:1702478727131:6efc2b0a)

jessicajaniuk commented:
This PR was merged into the repository by commit b06b24b5049c07fbc18c76fd2a10e49fc93870be.
